File: src/moddar/domain/link.py
class Link:
    """Represents a Reddit Link (post / submission)."""

    def __init__(
        self,
        id36: str,
        title: str,
        text: str,
        account_id: str,
        flairs: list[str] | None = None,
        url: str | None = None,
        images: list[str] | None = None,
    ) -> None:
        self._id36 = id36
        self._title = title
        self._text = text
        self._account_id = account_id
        self._flairs: list[str] = flairs if flairs is not None else []
        self._url: str | None = url
        self._images: list[str] | None = images

    @property
    def id36(self) -> str:
        return self._id36

    @property
    def title(self) -> str:
        return self._title

    @property
    def text(self) -> str:
        return self._text

    @property
    def account_id(self) -> str:
        return self._account_id

    @property
    def flairs(self) -> list[str]:
        return self._flairs

    @property
    def url(self) -> str | None:
        return self._url

    @property
    def images(self) -> list[str] | None:
        return self._images

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Link):
            return NotImplemented
        return (
            self._id36 == other._id36
            and self._title == other._title
            and self._text == other._text
            and self._account_id == other._account_id
            and self._flairs == other._flairs
            and self._url == other._url
            and self._images == other._images
        )

    def __repr__(self) -> str:
        return (
            f"Link(id36={self._id36!r}, title={self._title!r}, "
            f"text={self._text!r}, account_id={self._account_id!r}, "
            f"flairs={self._flairs!r}, url={self._url!r}, "
            f"images={self._images!r})"
        )

File: src/server/domain/link.ts
/** Represents a Reddit Link (post / submission). */
export class Link {
  private readonly _id36: string;
  private readonly _title: string;
  private readonly _text: string;
  private readonly _accountId: string;
  private readonly _subreddit: string | null;
  private readonly _flairs: readonly string[];
  private readonly _url: string | null;
  private readonly _images: readonly string[] | null;

  constructor({
    id36,
    title,
    text,
    accountId,
    subreddit = null,
    flairs = null,
    url = null,
    images = null,
  }: {
    id36: string;
    title: string;
    text: string;
    accountId: string;
    subreddit?: string | null;
    flairs?: readonly string[] | null;
    url?: string | null;
    images?: readonly string[] | null;
  }) {
    this._id36 = id36;
    this._title = title;
    this._text = text;
    this._accountId = accountId;
    this._subreddit = subreddit;
    this._flairs = flairs ?? [];
    this._url = url;
    this._images = images;
  }

  get id36(): string {
    return this._id36;
  }

  get title(): string {
    return this._title;
  }

  get text(): string {
    return this._text;
  }

  get accountId(): string {
    return this._accountId;
  }

  get subreddit(): string | null {
    return this._subreddit;
  }

  get flairs(): readonly string[] {
    return this._flairs;
  }

  get url(): string | null {
    return this._url;
  }

  get images(): readonly string[] | null {
    return this._images;
  }

  /** Value-based equality: compares all fields. */
  equals(other: unknown): boolean {
    if (!(other instanceof Link)) {
      return false;
    }
    return (
      this._id36 === other._id36 &&
      this._title === other._title &&
      this._text === other._text &&
      this._accountId === other._accountId &&
      this._subreddit === other._subreddit &&
      this._flairs.length === other._flairs.length &&
      this._flairs.every((f, i) => f === other._flairs[i]) &&
      this._url === other._url &&
      (this._images === null
        ? other._images === null
        : other._images !== null &&
          this._images.length === other._images.length &&
          this._images.every((img, i) => img === other._images![i]))
    );
  }

  toString(): string {
    return (
      `Link(id36=${JSON.stringify(this._id36)}, ` +
      `title=${JSON.stringify(this._title)}, ` +
      `text=${JSON.stringify(this._text)}, ` +
      `accountId=${JSON.stringify(this._accountId)}, ` +
      `subreddit=${JSON.stringify(this._subreddit)}, ` +
      `flairs=${JSON.stringify(this._flairs)}, ` +
      `url=${JSON.stringify(this._url)}, ` +
      `images=${JSON.stringify(this._images)})`
    );
  }
}
